File: 0037 - Sudoku Solver/Sudoku Solver.cpp
class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
        solve(board, 0, 0);
    }
    
    bool isValid(vector<vector<char>>& board, int i, int j, char ch) {
        for (int k = 0; k < 9; ++k) {
            if (board[i][k] == ch) {
                return false;
            }
        }
        for (int k = 0; k < 9; ++k) {
            if (board[k][j] == ch) {
                return false;
            }
        }
        for (int k = i / 3 * 3; k < 3 + i / 3 * 3; ++k) {
            for (int l = j / 3 * 3; l < 3 + j / 3 * 3; ++l) {
                if (board[k][l] == ch) {
                    return false;
                }
            }
        }
        return true;
    }
    
    bool solve(vector<vector<char>>& board, int i, int j) {
        if (j == 9) {
            j = 0;
            ++i;
        }
        if (i == 9) {
            return true;
        }
        if (board[i][j] != '.') {
            return solve(board, i, j + 1);
        }
        for (char c = '1'; c <= '9'; ++c) {
            if (isValid(board, i, j, c)) {
                board[i][j] = c;
                if (solve(board, i, j + 1)) {
                    return true;
                }
                board[i][j] = '.';
            }
        }
        return false;
    }
};

File: 0037 - Sudoku Solver/Sudoku Solver.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        Solution.solve(board, 0, 0)
        
    def solve(board, i, j):
        if j == 9:
            i = i + 1
            j = 0
        
        if i == 9:
            return True
        
        if board[i][j] != '.':
            return Solution.solve(board, i, j + 1)
        
        for c in ['1', '2', '3', '4', '5', '6', '7', '8', '9']:
            if Solution.isValid(board, i, j, c):
                board[i][j] = c
                if Solution.solve(board, i, j + 1):
                    return True
                board[i][j] = '.'
        return False
        
    def isValid(board, i, j, ch):
        for k in range(9):
            if board[k][j] == ch:
                return False
            
        for k in range(9):
            if board[i][k] == ch:
                return False
            
        for k in range(i // 3 * 3, 3 + i // 3 * 3):
            for l in range(j // 3 * 3, 3 + j // 3 * 3):
                if board[k][l] == ch:
                    return False
                
        return True

